roofit/roofitcore/src/RooAbsReal.cxx

```diff
@@ -4814,81 +4814,64 @@ void RooAbsReal::setParameterizeIntegral(const RooArgSet& paramVars)
 /// \return     Span pointing to the results. The memory is owned by `evalData`.
 RooSpan<double> RooAbsReal::evaluateSpan(RooBatchCompute::RunContext& evalData, const RooArgSet* normSet) const {
 
-  // Find leaves of the computation graph. Assign known data values to these.
-  //
-  // We can't use RooAbsArg::leafNodeServerList to find all leaves, sometimes a
-  // RooAbsReal sits on top of a leaf in the computation graph but it doesn't
-  // depend on it's values. The example here is a RooRealIntegral, which sets
-  // the leaf values itself to integrate over them. That's why we only add the
-  // parameters and observables here.
-  RooArgSet allLeafs;
-  RooArgSet parameters;
-  RooArgSet observables;
-  getParameters(normSet, parameters);
-  getObservables(normSet, observables);
-  allLeafs.add(parameters);
-  allLeafs.add(observables);
-
-  std::vector<RooAbsRealLValue*> settableLeaves;
-  std::vector<RooSpan<const double>> leafValues;
-  std::vector<double> oldLeafValues;
-
-  for (auto item : allLeafs) {
-    if (!item->IsA()->InheritsFrom(RooAbsRealLValue::Class()))
-      continue;
+  // Find all servers that are serving real numbers to us, retrieve their batch data,
+  // and switch them into "always clean" operating mode, so they return always the last-set value.
+  struct ServerData {
+    RooAbsReal* server;
+    RooSpan<const double> batch;
+    double oldValue;
+    RooAbsArg::OperMode oldOperMode;
+  };
+  std::vector<ServerData> ourServers;
+  std::size_t dataSize = 1;
+
+  for (auto absArgServer : servers()) {
+    if (absArgServer->IsA()->InheritsFrom(RooAbsReal::Class())) {
```

Comment:
```suggestion
    if (absArgServer->IsA()->InheritsFrom(RooAbsReal::Class()) && absArgServer->isValueServer(*this)) {
```
These test failures are because some integrals that are actually scalar are computed for each event if `RooAbsReal::evaluateSpan` is not doing something about it (see https://github.com/root-project/root/blob/master/roofit/roofitcore/src/RooAbsPdf.cxx#L372).

To avoid this, you should skip servers in this loop that are not value servers.

For example, if you have a Gaussian with scalar parameters `mu` and `sigma` and observable `x`, the observable is not a value server of the integral over x (since the integral sets the value itself). If you don't check for this, `dataSize` will be the size of the `x`-batch, and the integral gets computed for each event.

Reply:
👍